### ISSUE TITLE
Feedback round

### ISSUE THREAD

**Issue by kof** [andreipfeiffer/css-in-js#9]
Resulting from our conversation on Twitter, here is my Feedback. https://twitter.com/pfeiffer_andrei/status/1357304032059080704?s=20, this feedback mostly concerns with the common things across all libs and specific things for JSS. I haven't read all the specific info about other libs. 

Hope this helps to improve the document. I am sorry if anything sounds offensive. This is meant to be a constructive critique and as usual, on the internet, it's hard to convey a critique without offending.

### syntax highlighting

I am guessing you are talking about CSS specific syntax highlighting here, which is not clearly explained there. It seems you set on green the libraries that have template strings syntax support and have vscode plugins for syntax highlighting. The same libraries support object-based syntax and I am guessing they won't have syntax highlighting if you use it.

It needs to be clear that is basically saying there will be syntax highlighting when using template strings.
As a side note, JSS has an optional plugin for template strings, I am guessing syntax highlighting tools react on ` css`` `, so if one really wanted they could write a one-liner CSS tag that just returns the same string for JSS and it would highlight. I agree though if this doesn't come out of the box, people might not know they can do it. Maybe this should be added.

### code-completion for CSS properties and values

Here it's not clear what type of completion you are talking about. Is it based on typedefs or emmet, something else?
If you are talking about types based completion, there is a big difference between validating valid property names, existing CSS values OR it actually does completely check if a property X can be used with a value Y.

You will find a large discrepancy between different libraries, typescript, flowtype, and the 3 above named levels of guarantees.

## tag` `

As I already mentioned JSS provides it through an [optional plugin](https://cssinjs.org/jss-plugin-template), in a limited/minimal capacity. I find parsing CSS with JS in runtime is a bad idea if it's not done at build time. I am not sure this should go as a pure benefit, where it has bundle and runtime cost.

### "more suitable for new projects, when you don't need to migrate existing CSS"

This is not true, you can start using it at any point in time and migrate one component after another. If you need to reuse a CSS rule in the legacy part and in the new part at the same time during the migration period, you either make sure the rule is encapsulated inside a component and the same component is reused or you reuse it by composing globally scoped rules and local once in the new parts. For example, JSS [offers composition plugin](https://cssinjs.org/jss-plugin-compose), or you can add a second class to the element directly and there are some more ways to achieve that.

### without TS support, you won't get code completion

Or flowtype

## TS

JSS has TS defs built-in, I am not sure how good they are since I am not personally maintaining them, but they are definitely there.

## & ctx

In the case of JSS, I don't know why it says it's partial support, JSS had this syntax from day 1.

## Nesting

In the case of JSS, I don't know why it says it's partial support, JSS had this syntax from day 1.

## Thoughts on bundle size

It's hard to compare bundle size when a library is composed of plugins and they are all optional like it is in the case of JSS.
Because of plugins architecture, there are also different features, for e.g.  JSS core is 6.6kB  according to bundlephobia. What you mention in the table is react-jss, which contains some features other libraries don't have. It bundles the [default preset package](https://cssinjs.org/jss-preset-default), which bundles most official plugins with an exception of a few.

I am not sure which specific packages you have been using to get the bundle size data in case of all the other libs. This should be made somehow very clear, e.g. emotion is pretty much a collection of packages.

## Global styles

"JSS has a convoluted API for this, which requires an additional plugin, which we didn't figure out how to implement"

I have no idea why you say it because the API itself is straight forward and the plugin is part of the default preset, which is part of react-jss package I assume you have been trying. So all it should take is to use the [syntax](https://cssinjs.org/jss-plugin-global), essentially the same syntax as the rest of it, but inside of a @global rule, so keys become global selectors instead of rule names with generated selectors.

##  Vendor prefixes

"JSS requires an additional plugin for this"

Again if you tested react-jss, it's already part of it, since it's in the default preset.

## Unique class names

Actually, there is a big difference in algorithms used for that. Different hashing methods vs. counters based like JSS does it.

##  No inline styles

" pioneers like Radium & Glamor. " neither of them are pioneers in cssinjs and only Radium uses a hybrid approach with inline .styles

## FCP

```
For solutions that don't support .css file extraction, SSRed styles are added as <style> tags in the <head>, which will result in higher FCP than using regular CSS, because .css files can and will be loaded in parallel to other resources, while big <style> content will be sent and parsed along with the HTML, increasing parsing time.
```

Honestly, this is entirely wrong and I don't know where this misinformation comes from, sorry.

Critical CSS in the head are designed SPECIFICALLY to improve FCP. It allows the document to not have a render-blocking link tag with CSS in the head and start to render the contents as the HTML document arrives with correct styles. 
If you have a gigantic page and your critical CSS is the reason why the document is too slow before there is first contentful paint, you will have an even bigger problem with an extracted blocking CSS file when a user comes with an empty cache.

## Debugging / Inspecting

```
Most solutions inject the <style> tag in the DOM in DEVELOPMENT, which is a slower approach, but enables style inspecting using browser dev tools. But when building for PRODUCTION, they use CSSStyleSheet.insertRule() to inject the styles directly into the CSSOM, which is a way faster approach, but you cannot inspect the styles.

    JSS and Stitches use insertRule() in dev mode as well, so you cannot see what gets injected
    TypeStyle does NOT use insertRule(), not even in production
```
> Most solutions inject the <style> tag in the DOM in DEVELOPMENT, which is a slower approach,

Slower approach than what, link tag with a CSS file? Slower how/why?

> but enables style inspecting using browser dev tools

No, this is not what enables dev tools. There are a number of apis to render styles inside of style tag, style.textContent, style.innerHTML, style.appednChild(textNode) and finally style.sheet.insertRule()

>     JSS and Stitches use insertRule() in dev mode as well, so you cannot see what gets injected

JSS uses a hybrid approach, static styles are inserted using .textContent, dynamic once (function values) using insertRule

I believe emotion and styled components also use insertRule, maybe not always, needs verification

> which is a way faster approach, but you cannot inspect the styles.

No. Actually insertRule is slower than textContent, but marginally. What makes .textContent slow is the fact that you can't add  rules, when you set .textÇontent it sets the entire CSS string and the browser has to parse this entire string.

A one-of render with .textContent of a larger amount of CSS rules will be significantly faster than the same rules inserted using insertRule.

> but you cannot inspect the styles.

This is not true. What you can't do is see the styles in style element in elements tab in dev tools.
What you always could do is to see the styles in the styles panel in dev tools and it was always pointing to the style element they come from if you wanted to see it. In addition to that, you can always go into object inspector of the style element, open style.sheet.cssRules and see all the rules there.

The actual problem browsers had for no reason is that you couldn't EDIT the style rules inserted using insertRule API, you could always overwrite them with inline styles if you needed to try something out, in the same style tab, just couldn't edit the CSS rule itself.

This has changed a few months ago in chrome dev tools btw. You still can't see the styles in elements inspector but that's a spec design, insertRule is a CSSOM API, elements inspector works with the DOM API. When you use insertRule, you don't generate a DOM representation for those styles, it goes directly into CSSOM.

For example style.textContent is a DOM API that interacts with CSSOM internally and creates the CSS rules AND the dom string representation, so when this one is used, you can see CSS in the elements tab inside style element.

## Launched in 2016

Actually no, its launched in 2014

**Comment by andreipfeiffer** [comment]
Hi Oleg,

First of all, thank you so much for taking the time to send me this feedback. At the end there's a list of **todos** which include all the updates I will include, regarding JSS and the overall observations and clarifications.

Also, please notice the **3 questions at the end**.

<br />

### DX
You are right, the explanations are not very clear, as everything was written like **_"notes to myself"_**. This should be explained better.

### TS
The typings are definitely there, I've double checked them. I'm not a TS expert, and the types are pretty complex so I have no idea what the problem is, but they are **not working**. See for yourself:

https://user-images.githubusercontent.com/2570562/107009584-ed147780-679d-11eb-8f04-6e95173e9e6e.mp4

### `&` ctx
The [mq example](https://cssinjs.org/jss-syntax?v=v10.5.1#media-queries) doesn't show an example with contextual styles, and I understand why (because you need the nesting plugin for that). So, I finally figured out that you can do this:

```js
const useStyles = createUseStyles({
  button: {
    // ... default styles

    "@media (min-width: 640px)": {
      "&": { // <- had to figure this on my own
        // ... overwrite styles
      },
    },
  }
});
```

The fact that I had to figure this on my own, and all other libraries had such examples in the docs, should be a signal that the docs are not completely explicit.

### Nesting
I didn't see that the nesting plugin ships by default with `react-jss`, will update this

### Bundle size
For each solution, I've included the bare minimum to integrate with Next.js, based on their ["official" example](https://github.com/vercel/next.js/tree/canary/examples/with-react-jss), which afaict also include the minimum dependencies to render something on the screen.

### Global styles
Ok, I've managed to set global styles, thanks to [SOF](https://stackoverflow.com/questions/54201412/how-can-i-add-style-to-the-body-element-with-jss). However:
- your docs have no mention of `injectSheet` on [react-jss](https://cssinjs.org/react-jss?v=v10.5.1) page
- the [globals plugin](https://cssinjs.org/jss-plugin-global?v=v10.5.1) docs show only the definition example, not how to add the styles on the page
- the example on global page takes you to a 404
- I've searched the core api & syntax pages, absolutely no mention of `injectSheet`

So you see, I have no idea how anyone using your docs would figure out how to use global styles with React. The single fact that I wasn't able to find the information on my own is a hard signal that the docs are convoluted and incomplete. Where should I look for an example with "global styles, with React, with SSR"? on which of the separate pages? Without a search feature, anyone would simply give up.

### Vendor prefix
I didn't see that the plugin ships by default with `react-jss`, will update this

### FCP
My rationale is like this, so please correct me if I'm wrong:

- a SSRed page with `.css` extracted files would (typically) load like this:
   - get the html file
   - parse the header
   - start downloading the `.css` file **in parallel** with other `.js` files or other resources defined in head (a static `.css` file would normally be larger than the critical CSS included in the variant described below)
   - continue loading the rest of the page  

<br />

- a SSRed page with it's styles inside a `<style>` tag in the `<head>` of the `.html` page would (typically) load like this:
   - get the html file (which is larger, because of the `<style>` tag), so downloading it will take longer
   - parse the header (which contains global styles & critical CSS styles of the page, right? but these styles should be typically smaller than the `.css` file described in the above scenario) ... but parsing would take longer
   - start downloading any other `.js` files or other resources in the header
   - continue loading the rest of the page
   - at some point, load the runtime library for css-in-js
   - also, load the styles during hydration (which basically, are duplicates of the SSRed styles, and typically larger, because it contains all the styles, not only the critical ones)
   - parse these 2 files extra files

So, as far as I understand these steps:
1. TTFB is irrelevant, doesn't depend on the implementation
2. TTI is definitely longer, right?
3. so, the next relevant metric, which includes downloading and parsing of resources is FCP, depends a lot on the proportion of critical CSS styles and the total number of styles:
   - in an extreme scenario, where 99% of the styles are also critical CSS, the `.css` file approach has definitely a faster FCP
   - in the other extreme scenario, where 1% of the styles are critical CSS, the `<style>` tag should have a faster FCP

Of course, it's more complicated than that:
- it depends of the exact size (not the same comparing a 1kB or a 1MB of styles)
- it depends on other resources that you might have, and how and when you load them

However, I'm trying to figure out typical use-cases where one approach would out-perform the other (and vice-versa).

Curious on getting your opinion on this.

### Debugging / Inspecting
You're right, thank you for the explanation on how to see the content of the `<style>` tag added with `insertRule()`. Also, thank you the in-depth explanations, I will use your feedback to synthesise a clearer content.

<br />

# Todos

- [x] explain syntax highlighting more clear
- [x] explain code completion more clear
- [x] remove mention about `{}` syntax being more suitable for new projects
- [x] update mention about Radium/Glamour on the inline styles
- [x] update FCP explanation
- [x] update inspect/debug content
- [x] add optional tagged strings support for JSS
- [x] set launch date of JSS to 2014
- [x] add size & link for JSS nesting plugin (original size: 3.6kB / gzipped size: 1.4kB)
- [x] update nesting support for JSS to green
- [x] add a note to JSS regarding what ships with react-jss
- [x] update globals support for JSS
- [x] add a note about confusing, incomplete docs for JSS
- [x] update vendor prefix support for JSS to green
- [x] update bundle sizes for JSS

<br />

# Questions

1. Can you elaborate on the `Different hashing methods vs. counters` used for class names? Do you have an article that explain all methods? What are the tradeoffs? Do you think somebody would use this information to choose a library or not?

2. Regarding the loading metrics, would you agree that TTI is definitely faster with `.css` file extraction, and FCP depends a lot on the proportion between critical CSS and total CSS?

3. Is it true that if you want to support string templates syntax, you need to ship some additional code for runtime (babel-plugin, or something) to parse the code in order to inject the styles?

**Comment by kof** [comment]
Thanks for the pointers I will see what can be done about the docs, currently they suck.

> continue loading the rest of the page

HTML document is streaming, yes but it won't start rendering the body if blocking scripts/link tags are in the head until they are loaded and parsed/executed.

> start downloading any other .js files or other resources in the header

anything blocking in the head can completely ruin FCP performance, there should be nothing blocking in the head.

> a SSRed page with it's styles inside a <style> tag 
> continue loading the rest of the page

Not just loading, but rendering as it comes, not waiting for a response to fetch a ressource and execute it

> at some point, load the runtime library for css-in-js

After the FCP, so it doesn't influence how fast FCP is

> TTI is definitely longer, right?

Depends on how fast the JS can be executed, basically depends on js bundling and size of the actual code.
Compared to blocking CSS / JS, it will have the same cost, just at a different point of time, if you defer the execution of js and let the browser render HTML with critical CSS as it comes.

> in an extreme scenario, where 99% of the styles are also critical CSS, the .css file approach has definitely a faster FCP

If critical CSS has the same size as the full css file, first of all it's not critical and doesn't make sense any more. Second it will depend on the connection to the document vs. static server. If both have equal speed there is still latency to pay when downloading the external resource with empty cache.

> However, I'm trying to figure out typical use-cases where one approach would out-perform the other (and vice-versa).

If there is SSR and critical CSS and no blocking resources in the head and user has empty cache, I don't know of a case where external CSS resource would lead to a faster FCP.

Contra points are essentially SSR itself as it has a computational cost, requires all the frameworks stuff to work and be setup which is generally much more complex than simple HTML generator and some static CSS.

Also if it's a site with a very high chance of users coming with a full cache, then all optimisations are for minority. AFAIK full cache user is a minority for most cases, in particular when products are changing a lot.


> Can you elaborate on the Different hashing methods vs. counters used for class names? Do you have an article that explain all methods? What are the tradeoffs? Do you think somebody would use this information to choose a library or not?

Many use murmurhash, it requires to stringify the styles and then to generate a hash. This is much much slower than counters, basicaly incrementing a number on each new rule.

Pros Murmurhash
- results in the same hash for the same styles

Pros Counters
- much faster
- no library for hash creation needed (less payload)

> Regarding the loading metrics, would you agree that TTI is definitely faster with .css file extraction

Nope, TTI depends on 2 factors, FCP and execution or runtime code. FCP is faster with critical, execution runtime depends on your code and how you bundle the minimum you need, that's what react is trying to optimize with suspense and server components.


> Is it true that if you want to support string templates syntax, you need to ship some additional code for runtime (babel-plugin, or something) to parse the code in order to inject the styles?

Depends on the implementation, some libs avoid bundling CSS parser by providing a babel plugin and doing it at build time.

**Comment by kof** [comment]
One  more point I forgot: hydration has a cost for both components and cssinjs. The actual cost for rendering a more complete CSS as components hydrate will depend on the implementation of the CSSinJS library and the user styles. So for TTI, if we assumed user came with a full cache and external CSS ressource didn't block anything, it will be faster because js will only have to hydrate html.

**Comment by andreipfeiffer** [comment]
Hey Oleg,
Thank you again for the detailed arguments.

While I generally understand how page load works, there were some details I didn't take into consideration, which made **my mental model somewhat inaccurate**. I will make the necessary updates.

<br />

However, I have another quick curiosity regarding **Critical CSS extraction**.  

1. When you say "critical CSS", do you mean:
   a. CSS needed to style the entire static page (excluding dynamic/lazy loaded styles)?
   b. or "above the fold" CSS, which is what I initially thought, and what the web generally refers to ([web.dev](https://web.dev/extract-critical-css/), [SM](https://www.smashingmagazine.com/2015/08/understanding-critical-css/), [WPT](https://www.webperftools.com/blog/how-to-determine-critical-css-for-your-webpage/), [CSS-tricks](https://css-tricks.com/annotating-critical-css/), etc)

2. Do you know which CSS-in-JS libraries refer to bullet `a.` when they refer to critical CSS, and which refer to `b.` (if any)?

**Comment by kof** [comment]
Critical CSS is the minimal CSS required to paint the page on the initial load, before JS based, external CSS or anything else is available. It could be even less than the entire page. If you have lazy loading, you could style only viewport and delay the rest. In particular it is important that those minimal styles don't contain all possible dynamic states each element on the page can have once js is loaded, because this would bloat critical CSS. E.g. with css-modules and similar you can't have that, because it doesn't know what particular state for each component has been rendered during SSR.

Afaik emotion, SC, JSS all refer to 1.

**Comment by kof** [comment]
Actually with css-modules having critical CSS is potentially possible if one writes some abstraction around imported styles map, wraps each class with getters and aggregates only style rules which have been accessed during SSR.

**Comment by andreipfeiffer** [comment]
Ok, that makes sense.

**Comment by andreipfeiffer** [comment]
Hey Oleg,
I've finally managed to update the page based on your entire feedback.

Would mean a lot if you get a few minutes to look over, especially over the [Performance Metrics](https://github.com/andreipfeiffer/css-in-js/blob/main/README.md#-performance-metrics) and [Debugging / Inspecting](https://github.com/andreipfeiffer/css-in-js/blob/main/README.md#-debugging--inspecting) sections, and let me know if I;m missing something.

Best regards,
Andrei

**Comment by kof** [comment]
Looked over those 2 sections

> There are 2 methods to inject & update styles into the DOM from JavaScript:

I would say CSS instead of styles, because styles includes inline styles

> this method is apparently more performant than the previous one, so most libraries use this method in PRODUCTION;

Like I mentioned before it's only more performant for adding rules to existing once

Otherwise looks much better now!

**Comment by andreipfeiffer** [comment]
Hey @kof thanks for the corrections, will make the necessary updates.

**Comment by kof** [comment]
Btw. there is another approach, which is using `style.appendChild(document.createTextNode())` ... it allows using .textContent first for initial render and add more nodes later and it is visible in elements inspector but its much much slower than insertRule for adding nodes

**Comment by andreipfeiffer** [comment]
Got it. I think there's a lot to talk about if we were to dissect the rendering performance.

I don't plan to go very deep regarding this topic, as there are some studies in this regard already. I have a disclaimer that I don't touch upon the rendering performance, or "which library is faster".

Also, as [Max](https://www.youtube.com/watch?v=75kmPj_iUOA) is saying somewhere is his talk, nowadays all libraries have similar performance, because they use more or less the same methods.

I don't think that these implementation details play a huge role when somebody is looking to choose what library to use.
